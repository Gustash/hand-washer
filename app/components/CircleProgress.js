import React from 'react'
import { StyleSheet } from 'react-native'
import Svg, { Circle } from 'react-native-svg'
import Animated from 'react-native-reanimated'
import { colors } from 'app/config'

const { interpolate, multiply } = Animated
const AnimatedCircle = Animated.createAnimatedComponent(Circle)

const CircleProgress = ({ size, strokeWidth, progress }) => {
  const r = (size - strokeWidth) / 2
  const cx = size / 2
  const cy = size / 2
  const circumference = r * 2 * Math.PI

  // The current angle of the circle based on the progress.
  const angle = interpolate(progress, {
    inputRange: [0, 1],
    outputRange: [Math.PI * 2, 0],
  })
  const strokeDashoffset = multiply(angle, r)

  return (
    <Svg width={size} height={size} style={styles.container}>
      <AnimatedCircle
        stroke={colors.green}
        fill="none"
        strokeDasharray={`${circumference}, ${circumference}`}
        strokeDashoffset={strokeDashoffset}
        strokeWidth={strokeWidth}
        cx={cx}
        cy={cy}
        r={r}
        strokeLinecap="round"
      />
    </Svg>
  )
}

const styles = StyleSheet.create({
  container: {
    transform: [{ rotateZ: '270deg' }],
  },
})

export default CircleProgress
