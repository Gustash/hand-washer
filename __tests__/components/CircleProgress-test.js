import React from 'react'
import Animated from 'react-native-reanimated'
import { Circle } from 'react-native-svg'
import { shallow } from 'enzyme'
import { shallowToJson } from 'enzyme-to-json'

import CircleProgress from '../../app/components/CircleProgress'

const { Value } = Animated
const AnimatedCircle = Animated.createAnimatedComponent(Circle)

const mockProps = props => ({
  size: 24,
  strokeWidth: 4,
  progress: new Value(0),
  ...props,
})

describe('<CircleProgress />', () => {
  const testProps = mockProps({
    size: 64,
    strokeWidth: 9,
    progress: new Value(0.5),
  })
  let initialProps
  let wrapper

  beforeEach(() => {
    initialProps = mockProps()
    wrapper = shallow(<CircleProgress {...initialProps} />)
  })

  it('matches CircleProgress snapshot', () => {
    expect(shallowToJson(wrapper)).toMatchSnapshot()
  })

  it('calculates radius', () => {
    expect(wrapper.find(AnimatedCircle).prop('r')).toBe(10)
    wrapper.setProps(testProps)
    expect(wrapper.find(AnimatedCircle).prop('r')).toBe(27.5)
  })

  it('calculates cx', () => {
    expect(wrapper.find(AnimatedCircle).prop('cx')).toBe(12)
    wrapper.setProps(testProps)
    expect(wrapper.find(AnimatedCircle).prop('cx')).toBe(32)
  })

  it('calculates cy', () => {
    expect(wrapper.find(AnimatedCircle).prop('cy')).toBe(12)
    wrapper.setProps(testProps)
    expect(wrapper.find(AnimatedCircle).prop('cy')).toBe(32)
  })

  it('calculates circumference', () => {
    expect(wrapper.find(AnimatedCircle).prop('strokeDasharray')).toBe(
      '62.83185307179586, 62.83185307179586'
    )
    wrapper.setProps(testProps)
    expect(wrapper.find(AnimatedCircle).prop('strokeDasharray')).toBe(
      '172.78759594743863, 172.78759594743863'
    )
  })
})
