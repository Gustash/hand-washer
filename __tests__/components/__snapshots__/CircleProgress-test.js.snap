// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CircleProgress /> matches CircleProgress snapshot 1`] = `
<Svg
  height={24}
  preserveAspectRatio="xMidYMid meet"
  style={
    Object {
      "transform": Array [
        Object {
          "rotateZ": "270deg",
        },
      ],
    }
  }
  width={24}
>
  <Circle
    cx={12}
    cy={12}
    fill="none"
    r={10}
    stroke="#00A778"
    strokeDasharray="62.83185307179586, 62.83185307179586"
    strokeDashoffset={
      AnimatedValue {
        " __value": NaN,
      }
    }
    strokeLinecap="round"
    strokeWidth={4}
  />
</Svg>
`;
